src/middlewared/middlewared/plugins/snmp.py

```diff
@@ -1,10 +1,126 @@
+from middlewared.utils.crypto import generate_string
 from middlewared.validators import Email, Match, Or, Range
 
+SNMP_PRIV_CONF = '/var/lib/snmp/snmpd.conf'
+SNMP_PRIV_KEY = '/var/lib/private/snmp'
+_SNMP_SYSTEM_USER = {
+    'name': 'snmpSystemUser', 'auth_type': 'SHA', 'key': None, 'size': 0
+}
+
+
+def _get_authuser_secret():
+    """
+    Get the auth user saved secret.
+    Internal helper function for use by this module.
+    Return decoded string.
+    """
+    secret = ""
+    if not _SNMP_SYSTEM_USER['key']:
+        # No system user key registered
+        return secret
+
+    try:
+        with open(SNMP_PRIV_KEY, 'rb') as fd:
+            secret_c = fd.read(_SNMP_SYSTEM_USER['size'])
+    except FileNotFoundError:
+        return secret
+
+    try:
+        secret = Fernet(_SNMP_SYSTEM_USER['key']).decrypt(secret_c).decode()
+    except TypeError:
+        return secret
+
+    return secret
+
+
+def _set_authuser_secret(secret):
+    """
+    Save the auth user secret.
+    Internal helper function for use by this module.
+    INPUT: ascii string (not encoded)
+    """
+    secret_c = Fernet(_SNMP_SYSTEM_USER['key']).encrypt(secret.encode())
+    with open(SNMP_PRIV_KEY, 'wb') as fd:
+        _SNMP_SYSTEM_USER['size'] = fd.write(secret_c)
+
+    return
+
+
+def add_snmp_user(snmp):
+    """
+    Build the createUser message and add it to the private config file.
+    NOTE: The SNMP daemon should be stopped before calling this routine and
+            the new user will be available after starting SNMP.
+    """
+    # The private config file must exist, i.e. SNMP must have been started at least once
+    if not os.path.exists(SNMP_PRIV_CONF):
+        return
+
+    # Build the 'createUser' message
+    create_v3_user = f"createUser {snmp['v3_username']} "
+
+    user_pwd = snmp['v3_password']
+    create_v3_user += f'{snmp["v3_authtype"]} "{user_pwd}" '
+
+    if snmp.get('v3_privproto'):
+        user_phrase = snmp['v3_privpassphrase']
+        create_v3_user += f'{snmp["v3_privproto"]} "{user_phrase}" '
+
+    create_v3_user += '\n'
+
+    # Example: createUser newPrivUser MD5 "abcd1234" DES "abcd1234"
+    with open(SNMP_PRIV_CONF, 'a') as f:
+        f.write(create_v3_user)
+
+
+def delete_snmp_user(user):
+    """
+    Delete the SNMPv3 user
+    RETURN: stdout message
+    NOTE: SNMP must be running for this call to succeed
+    """
+    if pwd := _get_authuser_secret():
+        # snmpusm -v3 -l authPriv -u JoeUser -a MD5 -A "abcd1234" -x AES -X "A pass phrase" localhost delete JoeUser
+        cmd = [
+            'snmpusm', '-v3', '-u', f'{_SNMP_SYSTEM_USER["name"]}',
+            '-l', 'authNoPriv', '-a', f'{_SNMP_SYSTEM_USER["auth_type"]}', '-A', f'{pwd}',
+            'localhost', 'delete', user
+        ]
+        # This call will timeout if SNMP is not running
+        subprocess.run(cmd, capture_output=True)
+    else:
+        raise CallError
+
+
+def _add_system_user():
```

Comment:
It would probably be better to create a directory `snmp_` and put these into a `utils.py` file within it.

Reply:
agreed